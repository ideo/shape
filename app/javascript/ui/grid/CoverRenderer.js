import PropTypes from 'prop-types'
import { PropTypes as MobxPropTypes } from 'mobx-react'

import { routingStore } from '~/stores'
import LinkItemCover from '~/ui/grid/covers/LinkItemCover'
import TextItemCover from '~/ui/grid/covers/TextItemCover'
import PdfFileItemCover from '~/ui/grid/covers/PdfFileItemCover'
import ImageItemCover from '~/ui/grid/covers/ImageItemCover'
import VideoItemCover from '~/ui/grid/covers/VideoItemCover'
import GenericFileItemCover from '~/ui/grid/covers/GenericFileItemCover'
import CollectionCover from '~/ui/grid/covers/CollectionCover'
import DataItemCover from '~/ui/grid/covers/DataItemCover'
import ChartItemCover from '~/ui/grid/covers/ChartItemCover'
import LegendItemCover from '~/ui/grid/covers/LegendItemCover'

import { ITEM_TYPES } from '~/utils/variables'

class CoverRenderer extends React.Component {
  get isItem() {
    return this.props.cardType === 'items'
  }

  get isCollection() {
    return this.props.cardType === 'collections'
  }

  get renderCover() {
    const {
      card,
      dragging,
      record,
      height,
      handleClick,
      searchResult,
      isBoardCollection,
    } = this.props
    if (this.isItem) {
      switch (record.type) {
        case ITEM_TYPES.TEXT:
          return (
            <TextItemCover
              item={record}
              height={height}
              dragging={dragging}
              cardId={card.id}
              handleClick={handleClick}
              searchResult={searchResult}
              initialFontTag={isBoardCollection ? 'H1' : 'P'}
            />
          )
        case ITEM_TYPES.EXTERNAL_IMAGE:
          return <ImageItemCover item={record} contain={card.image_contain} />
        case ITEM_TYPES.FILE: {
          if (record.isPdfFile) {
            return <PdfFileItemCover item={record} />
          } else if (record.isImage) {
            return <ImageItemCover item={record} contain={card.image_contain} />
          } else if (record.isVideo) {
            return <VideoItemCover item={record} dragging={dragging} />
          } else if (record.filestack_file) {
            return <GenericFileItemCover item={record} />
          }
          return <div style={{ padding: '20px' }}>File not found.</div>
        }
        case ITEM_TYPES.VIDEO:
          return <VideoItemCover item={record} dragging={dragging} />
        case ITEM_TYPES.LINK:
          return (
            <LinkItemCover
              item={record}
              cardHeight={card.height}
              dragging={dragging}
            />
          )
        case ITEM_TYPES.CHART:
          return <ChartItemCover item={record} testCollection={card.parent} />

        case ITEM_TYPES.DATA:
          // We must pass in dataset length to trigger
          // re-render when new datasets are added
          return (
            <DataItemCover
              datasetLength={record.datasets.length}
              height={height}
              item={record}
              card={card}
            />
          )

        case ITEM_TYPES.LEGEND:
          return <LegendItemCover item={record} card={card} />

        default:
          return <div>{record.content}</div>
      }
    } else if (this.isCollection) {
      return (
        <CollectionCover
          width={card.maxWidth}
          height={card.maxHeight}
          collection={record}
          dragging={dragging}
          inSubmissionsCollection={
            card.parentCollection &&
            card.parentCollection.isSubmissionsCollection
          }
        />
      )
    }
    return <div />
  }

  handleClickToCollection = () => {
    const { card } = this.props
    routingStore.routeTo('collections', card.record.id)
  }

  render() {
    const { coverItem } = this.props
    if (coverItem) {
      return (
        <div
          onClick={this.handleClickToCollection}
          onKeyDown={this.handleClickToCollection}
          role="link"
          tabIndex="0"
        >
          {this.renderCover}
        </div>
      )
    }
    return this.renderCover
  }
}

CoverRenderer.propTypes = {
  card: MobxPropTypes.objectOrObservableObject.isRequired,
  cardType: PropTypes.string.isRequired,
  record: MobxPropTypes.objectOrObservableObject.isRequired,
  isBoardCollection: PropTypes.bool,
  coverItem: PropTypes.bool,
  height: PropTypes.number,
  dragging: PropTypes.bool,
  searchResult: PropTypes.bool,
  handleClick: PropTypes.func,
}

CoverRenderer.defaultProps = {
  height: 1,
  dragging: false,
  searchResult: false,
  coverItem: false,
  isBoardCollection: false,
  handleClick: () => null,
}

export default CoverRenderer
