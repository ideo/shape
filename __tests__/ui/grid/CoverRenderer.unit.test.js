import CoverRenderer from '~/ui/grid/CoverRenderer'
import expectTreeToMatchSnapshot from '#/helpers/expectTreeToMatchSnapshot'
import {
  fakeCollectionCard,
  fakeItemCard,
  fakeCollection,
  fakeTextItem,
  fakeFileItem,
} from '#/mocks/data'

let wrapper, rerender, props

describe('CoverRenderer', () => {
  beforeEach(() => {
    props = {}
    props.card = fakeItemCard
    props.cardType = 'items'
    props.record = fakeTextItem
    props.height = 1
    rerender = function() {
      wrapper = shallow(<CoverRenderer {...props} />)
      return wrapper
    }
  })

  it('renders snapshot', () => {
    expectTreeToMatchSnapshot(wrapper)
  })

  describe('with collection', () => {
    beforeEach(() => {
      props.card = fakeCollectionCard
      props.cardType = 'collections'
      props.canEditCollection = false
      props.record = fakeCollection
      props.record.can_edit = false
      rerender()
    })

    it('renders the collection cover', () => {
      expect(wrapper.find('CollectionCover').props().collection).toEqual(
        fakeCollection
      )
    })
  })

  describe('with a pdf file', () => {
    beforeEach(() => {
      props.card.record = { ...fakeFileItem }
      props.record = props.card.record
      props.record.isPdfFile = true
      props.cardType = 'items'
      rerender()
    })

    it('renders a generic file cover', () => {
      expect(wrapper.find('PdfFileItemCover').exists()).toBeTruthy()
    })
  })

  describe('with a generic file', () => {
    beforeEach(() => {
      props.card.record = { ...fakeFileItem }
      props.record = props.card.record
      props.record.isGenericFile = true
      props.cardType = 'items'
      rerender()
    })

    it('renders a generic file cover', () => {
      expect(wrapper.find('GenericFileItemCover').exists()).toBeTruthy()
    })
  })
})
