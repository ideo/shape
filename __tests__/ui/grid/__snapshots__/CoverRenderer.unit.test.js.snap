// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoverRenderer renders snapshot 1`] = `undefined`;
